### PR TITLE
feat: add custom events to Relay

### LINKED ISSUE #1142: Relay interface doesn't include 'observers'
This is a **feature or change request**

## Problem

waku.relay (IRelay) does not provide public accessor for `observers`.

## Proposed Solutions

Make `observers` Map publicly accessible on IRelay interface.

## Notes

Current workaround: 

```
// @ts-ignore - private accessor
return waku.relay.observers.keys() as string[];
```

### PR DESCRIPTION
**Pull request by weboko**
## Problem
There was a [report](https://github.com/waku-org/js-waku/issues/1142#issuecomment-1444809317) that `Relay.observers` were not properly cleaned up and it was hard to debug in a runtime so some hack was introduced on the consumer side. 

## Solution

Add `observer:added` and `observer:removed` to track state of the internal `observers` structure.

## Notes

- Resolves #1142 

## Validation

image

### CODE REVIEW

**Review by fryorcraken — COMMENTED** [review]
Looks good. I only wonder if we should use `CustomEvents` from libp2p or have our own. The event does not seem libp2p related so just wonder if we are restricting ourselves here.

**Comment by weboko** [comment]
> Looks good. I only wonder if we should use `CustomEvents` from libp2p or have our own. The event does not seem libp2p related so just wonder if we are restricting ourselves here.

Not really, `CustomEvent` is just a browser [abstract](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent) and we can avoid using it altogether any time even now. 
Used it just to be consistent with `GossipSub`

File: packages/core/src/lib/relay/index.ts [hunk 1]
```diff
@@ -6,7 +6,8 @@ import {
 } from "@chainsafe/libp2p-gossipsub";
 import type { PeerIdStr, TopicStr } from "@chainsafe/libp2p-gossipsub/types";
 import { SignaturePolicy } from "@chainsafe/libp2p-gossipsub/types";
-import type {
+import type { CustomEvent } from "@libp2p/interfaces/events";
+import {
```
Comment (fryorcraken) [inline packages/core/src/lib/relay/index.ts:10]:
```suggestion
import type {
```

File: packages/core/src/lib/relay/index.ts
Comment (fryorcraken) [inline packages/core/src/lib/relay/index.ts:1]: Should we make `observers` property `private` now?

**Review by fryorcraken — APPROVED** [review]
Looks good bug I wonder if it's good enough for @therealjmj. We might need a dump function too.

File: packages/core/src/lib/relay/index.ts
Comment (weboko) [reply packages/core/src/lib/relay/index.ts:1]: good point

**Comment by therealjmj** [comment]
> Looks good bug I wonder if it's good enough for @therealjmj. We might need a dump function too.

Funny, in making 'observers' private, it's likely this broke our implementation.
We just want a way to log the current observers, as you say, a "dump".

**Comment by weboko** [comment]
@therealjmj I can followup with a "dump" function but isn't it enough to have these two events for you to track what happens to `observers`? If you have a need for full observers list it can be maintained by these events too.

**Comment by therealjmj** [comment]
> @therealjmj I can followup with a "dump" function but isn't it enough to have these two events for you to track what happens to `observers`? If you have a need for full observers list it can be maintained by these events too. 

I'm really not a fan of events - it'll add tech that needs to be tested and maintained on our end.

I initially wanted to be helpful, but now it's added work on our end 😅

**Comment by weboko** [comment]
Ok :) Adding dump function soon

**Pull request merged**